### PR TITLE
chore(template): service connect CFN changes

### PR DESCRIPTION
**Pull request by iamhopaul123**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: internal/pkg/template/templates/workloads/partials/cf/service-base-properties.yml [hunk 1]
```diff
@@ -33,6 +33,28 @@ CapacityProviderStrategy:
     {{- end}}
   {{- end}}
 {{- end }}
+{{- if .ServiceConnect }}
+ServiceConnectConfiguration:
+  Enabled: True
+  Namespace: {{.ServiceConnect.Namespace}}
```
Comment (paragbhingre) [inline internal/pkg/template/templates/workloads/partials/cf/service-base-properties.yml:39]: Is there any way Namespace can be nil?

Comment (iamhopaul123) [reply internal/pkg/template/templates/workloads/partials/cf/service-base-properties.yml:39]: Not really https://github.com/aws/copilot-cli/pull/4036/files#diff-53f2b158bad1d10014e86bdd2b5b6ccc88acb52a641ccfce4a65c9fad47517cdR327

**Review by huanjani — APPROVED** [review]
Exciting!

File: internal/pkg/template/templates/workloads/partials/cf/workload-container.yml [hunk 2]
```diff
@@ -38,8 +41,12 @@
 {{- end}}
 {{- end}} {{/* end if eq .WorkloadType "Load Balanced Web Service"*/}}
 {{- if eq .WorkloadType "Backend Service"}}
+{{- if .ServiceConnect }} # remove when release
+  PortMappings: !If [ExposePort, [{ContainerPort: !Ref ContainerPort, Name: !Ref WorkloadName}], !Ref "AWS::NoValue"]
```
Comment (Lou1415926) [inline internal/pkg/template/templates/workloads/partials/cf/workload-container.yml:47]: hmm I wonder if this works well with people that has task def overrides in their manifest

Comment (iamhopaul123) [reply internal/pkg/template/templates/workloads/partials/cf/workload-container.yml:47]: Possible if they override the whole portmapping but should be fine. it's not something we can guarantee.

File: internal/pkg/deploy/cloudformation/stack/lb_web_svc.go [hunk 3]
```diff
@@ -190,7 +190,7 @@ func (s *LoadBalancedWebService) Template() (string, error) {
 	if s.manifest.RoutingRule.RedirectToHTTPS != nil {
 		httpRedirect = aws.BoolValue(s.manifest.RoutingRule.RedirectToHTTPS)
 	}
-	_, targetContainerPort := s.httpLoadBalancerTarget()
+	targetContainer, targetContainerPort := s.httpLoadBalancerTarget()
```
Comment (dannyrandall) [inline internal/pkg/deploy/cloudformation/stack/lb_web_svc.go:193]: why was this removed initially..?

File: internal/pkg/template/templates/workloads/partials/cf/service-base-properties.yml [hunk 4]
```diff
@@ -33,6 +33,28 @@ CapacityProviderStrategy:
     {{- end}}
   {{- end}}
 {{- end }}
+{{- if .ServiceConnect }}
+ServiceConnectConfiguration:
```
Comment (dannyrandall) [inline internal/pkg/template/templates/workloads/partials/cf/service-base-properties.yml:37]: how does SC work with SD?

File: internal/pkg/template/templates/workloads/partials/cf/service-base-properties.yml [hunk 5]
```diff
@@ -33,6 +33,28 @@ CapacityProviderStrategy:
...
+  Enabled: True
+  Namespace: {{.ServiceConnect.Namespace}}
+  LogConfiguration:
+    LogDriver: awslogs
+    Options:
+      awslogs-region: !Ref AWS::Region
+      awslogs-group: !Ref LogGroup
```
Comment (dannyrandall) [inline internal/pkg/template/templates/workloads/partials/cf/service-base-properties.yml:44]: does this mean sc logs will get interleaved with container logs? or will they be separate? what kind of stuff is logged?

File: internal/pkg/template/templates/workloads/partials/cf/service-base-properties.yml [hunk 6]
```diff
@@ -33,6 +33,28 @@ CapacityProviderStrategy:
...
+      awslogs-stream-prefix: copilot
+  Services:
+    - PortName: TargetPort
+      # Avoid using the same service with Service Discovery in a namespace.
+      DiscoveryName: !Join ["-", [!Ref WorkloadName, "sc"]] 
+      ClientAliases:
+        - Port: !Ref TargetPort
```
Comment (dannyrandall) [inline internal/pkg/template/templates/workloads/partials/cf/service-base-properties.yml:51]: nice 👏

File: internal/pkg/template/templates/workloads/partials/cf/sidecars.yml [hunk 7]
```diff
@@ -47,6 +47,13 @@
 {{include "image-overrides" . | indent 2}}  
 {{- if $sidecar.Port}}
   PortMappings:
+{{- if $.ServiceConnect }} # remove when release
+{{- if eq $.HTTPTargetContainer.Container $sidecar.Name}}
+{{- if eq $.HTTPTargetContainer.Port $sidecar.Port}}
+      Name: TargetPort
```
Comment (dannyrandall) [inline internal/pkg/template/templates/workloads/partials/cf/sidecars.yml]: shouldn't the array mark be here? `-`? or just move this to the bottom of the object?

**Review by dannyrandall — COMMENTED** [review]
just some questions 😊

File: internal/pkg/deploy/cloudformation/stack/lb_web_svc.go (hunk 3 above)
Comment (iamhopaul123) [reply internal/pkg/deploy/cloudformation/stack/lb_web_svc.go:193]: no idea...not getting used?

File: internal/pkg/template/templates/workloads/partials/cf/service-base-properties.yml (hunk 4 above)
Comment (iamhopaul123) [reply internal/pkg/template/templates/workloads/partials/cf/service-base-properties.yml:37]: It's magic ✡️ basically if both SC and SD use the same dns name, and both client and service enable SC, they'll use SC otherwise they'll use SD.

File: internal/pkg/template/templates/workloads/partials/cf/service-base-properties.yml (hunk 5 above)
Comment (iamhopaul123) [reply internal/pkg/template/templates/workloads/partials/cf/service-base-properties.yml:44]: it'll create a separate logstream in the same log group. For example:

```
copilot/ecs-service-conne 'x-content-type-options', 'nosniff'
copilot/ecs-service-conne 'date', 'Tue, 27 Sep 2022 21:26:30 GMT'
copilot/ecs-service-conne 'server', 'envoy'
copilot/ecs-service-conne 'connection', 'close'
copilot/ecs-service-conne [2022-09-27 21:26:30.160][22][debug][misc] [source/server/admin/admin_filter.cc:82] nextChunk: response.length=67754 more_data=false end_stream=true
copilot/ecs-service-conne [2022-09-27 21:26:30.160][22][debug][connection] [source/common/network/connection_impl.cc:139] [C86665] closing data_to_write=67987 type=0
copilot/ecs-service-conne [2022-09-27 21:26:30.160][22][debug][connection] [source/common/network/connection_impl.cc:139] [C86665] closing data_to_write=67987 type=2
copilot/ecs-service-conne [2022-09-27 21:26:30.160][22][debug][connection] [source/common/network/connection_impl.cc:716] [C86665] write flush complete
copilot/ecs-service-conne [2022-09-27 21:26:30.160][22][debug][connection] [source/common/network/connection_impl.cc:250] [C86665] closing socket: 1
copilot/ecs-service-conne [2022-09-27 21:26:30.164][22][debug][conn_handler] [source/server/active_stream_listener_base.cc:120] [C86665] adding to cleanup list
```

File: internal/pkg/template/templates/workloads/partials/cf/sidecars.yml (hunk 7 above)
Comment (iamhopaul123) [reply internal/pkg/template/templates/workloads/partials/cf/sidecars.yml]: i was just dumb. Good find!

File: internal/pkg/template/workload.go [hunk 8]
```diff
@@ -210,7 +210,8 @@ type LogConfigOpts struct {
 
 // HTTPTargetContainer represents the target group of a load balancer that points to a container.
 type HTTPTargetContainer struct {
-	Port string // Port of the container.
+	Container string
```
Comment (efekarakus) [inline internal/pkg/template/workload.go:213]: can we rename the field to `Name` since the type has the word `Container` in it 🥺

**Review by dannyrandall — APPROVED** [review]
ty!🦃 looks great!
